src/DependencyInjection/Compiler/TemplateRegistryCompilerPass.php

```diff
@@ -0,0 +1,195 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\DependencyInjection\Compiler;
+
+use Sonata\AdminBundle\Admin\AbstractAdmin;
+use Sonata\AdminBundle\Templating\MutableTemplateRegistry;
+use Sonata\AdminBundle\Templating\MutableTemplateRegistryAwareInterface;
+use Sonata\AdminBundle\Templating\MutableTemplateRegistryInterface;
+use Sonata\AdminBundle\Templating\TemplateRegistry;
+use Sonata\AdminBundle\Templating\TemplateRegistryAwareInterface;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * @author Wojciech Błoszyk <wbloszyk@gmail.com>
+ */
+final class TemplateRegistryCompilerPass implements CompilerPassInterface
+{
+    const TEMPLATE_REGISTRY_TAG = 'sonata.admin.template_registry';
+    const MUTABLE_TEMPLATE_REGISTRY_TAG = 'sonata.admin.mutable_template_registry';
+    const TEMPLATE_REGISTRY_AWARE_TAG = 'sonata.admin.template_registry_aware';
+    const MUTABLE_TEMPLATE_REGISTRY_AWARE_TAG = 'sonata.admin.mutable_template_registry_aware';
+
+    public function process(ContainerBuilder $container): void
+    {
+        $this->templateRegistryProcess($container, self::TEMPLATE_REGISTRY_TAG);
+        $this->templateRegistryProcess($container, self::TEMPLATE_REGISTRY_AWARE_TAG);
+
+        $this->templateRegistryProcess($container, self::MUTABLE_TEMPLATE_REGISTRY_TAG);
+        $this->templateRegistryProcess($container, self::MUTABLE_TEMPLATE_REGISTRY_AWARE_TAG);
+    }
+
+    private function templateRegistryProcess(ContainerBuilder $container, string $tag): void
+    {
+        $taggedServices = $container->findTaggedServiceIds($tag);
+
+        foreach ($taggedServices as $id => $tags) {
+            $templates = [];
+
+            foreach ($tags as $attributes) {
+                if (isset($attributes['type'], $attributes['template'])) {
```

Comment:
IMO, the name "type" for this option is too restrictive. Shouldn't this be "template_name"?
My concern is about scenarios where the passed template could be considered something more general than the "type" concept.

Reply:
`template_name` seems better indeed.

Reply:
`sonata.admin.template_registry.template` tag with `name` and `value` arguments.

Reply:
Isn't `name` reserved ?

I mean, you currently used
```
- {name: sonata.admin.mutable_template_registry_aware, type:'edit', template: '@@FooAdmin/CRUD/edit.html.twig'}
```
So how could we use `name` instead of `type` since `name` is already used ?

Reply:
My bad. `template_name` and `template_value` to be consider. I will finish it in friday evening or in saturday.

Reply:
I'd use other name than `template_value`. Maybe `template_path` or just `template`.

Reply:
`template_path` will be the best.